pandas/core/indexes/interval.py

```diff
@@ -906,35 +906,35 @@ def get_indexer(
             )
             raise InvalidIndexError(msg)
 
-        target = ensure_index(target)
+        target_as_index = ensure_index(target)
```

Comment:
if you also make `AnyArrayLike` an Alias instead of a TypeVar, then this module doesn't need to be changed at all.

Reply:
Same thing though - this just loosens the type checking which isn't desired. Actually moved towards TypeVars for reasons described in https://github.com/pandas-dev/pandas/issues/26453#issuecomment-497025873

Might update #27050 to include some of that info

Reply:
> Actually moved towards TypeVars for reasons described in [#26453 (comment)](https://github.com/pandas-dev/pandas/issues/26453#issuecomment-497025873)

i don't think that applies here.

`target` does not need to be a TypeVar, the type of `target` does not need to be maintained.

`target` can be any of `["ExtensionArray", "Index", "Series", "SparseSeries", np.ndarray]`

`ensure_index` is not yet typed. but presumably will have a return type of `Index`.

so reassigning with an `Index` to a variable that has a type of Union of `["ExtensionArray", "Index", "Series", "SparseSeries", np.ndarray]` is perfectly valid.

Reply:
We could add Union alternatives for each TypeVar in the central module but I think that confounds the point of generic programming and/or makes our type system weaker. Another option would be to allow redefinition of variables which mypy supplies a setting for:

https://mypy.readthedocs.io/en/latest/command_line.html?highlight=allow-redefinition#miscellaneous-strictness-flags

But I also think that makes for a weaker type system, and generally there's not a lot of downside to creating a separate variable here instead of allowing it's type to implicitly be altered by the return of `ensure_index`

Reply:
> Another option would be to allow redefinition of variables which mypy supplies a setting for:
> 
> https://mypy.readthedocs.io/en/latest/command_line.html?highlight=allow-redefinition#miscellaneous-strictness-flags
> 
> But I also think that makes for a weaker type system

definitely. we should rule this out.


> generally there's not a lot of downside to creating a separate variable here instead of allowing it's type to implicitly be altered by the return of `ensure_index`

disagree. the return type of `ensure_index` would need to conform to the type of `target`.

is creating new variables not weakening the type system?